Monitor layout

fsih opened the pull request:
### Resolves
https://github.com/LLK/scratch-gui/issues/458 Monitors overlap each other on stage
https://github.com/LLK/scratch-gui/issues/472 Variable Monitor Cannot Be Moved
https://github.com/LLK/scratch-gui/issues/463 Monitor list extends outside of stage

### Proposed Changes

Algorithm for layout out monitors
Requirements:
- Monitors line up on the left edge of the screen first
- Monitors try to appear in a place where they won't overlap already showing monitors
- If a monitor is moved, then turned off and on again, it stays in the place it was put

Beyond that, I didn't try to match very closely the behavior of 2.0, which can be pretty strange. 
I tried to document some of it in https://github.com/LLK/scratch-gui/issues/463. You're pretty unlikely to want your entire project covered in monitors anyway.

Here is how I chose to implement this:
1. Monitors that were moved by the user always appear wherever they were last seen.
2. New monitors appear in the first leftmost, then topmost place they can fit without overlapping another monitor
    - If you make a variable, then hide its monitor, then make a new variable, the monitor for the new variable will appear in the very top left instead of below the space left open for the first monitor. This is different from how Scratch 2.0 does it and might take some iterating.
3. When the first column fills up, monitors start to line up one column to the right (avoiding rows where they would overlap the first column). A column is somewhat arbitrarily 50 pixels. (I tried having them go to the leftmost available spot, which was the best packing, but made them appear in a really weird order when you got to the second column.)
4. If the monitor doesn't fit anywhere, put it in the leftmost topmost place where it fits except that it goes off the screen to the left or bottom. Also, only do this if it's still at least 40 pixels on-screen. This is so that if you name your variable `asdfasdfasdfadsfafdafdsadfasdfasfsadfsafsafsdfsafdasfdsafdsafaasdfsafadsfaasfdasfdasfdsadfasdfsafdsadfasdfasfasdf`, it still tries to line it up somewhere.
5. If the entire screen is covered in monitors, place the monitor randomly on the left half of the screen.
![image](https://user-images.githubusercontent.com/2855464/33453541-f7214086-d5e3-11e7-8167-08d6a9efcd11.png)
(4 and 5 are overlapping because the monitor was moved on top of them after they spawned. Notice how 11 avoided the moved monitor)

### Reason for Changes

Fix bugs

fsih commented:
poke

paulkaplan commented:
@fsih I'm concerned about how much this accesses the dom directly to write and read style properties, instead of using the APIs provided by react/react-draggable. There can be considerable performance penalties for doing things like reading and writing css properties from the dom during a drag. Want to just talk through that part when you are in? Specifically parts like this https://github.com/LLK/scratch-gui/pull/976/files#diff-352d73d98a71b1989c638208d9888b84R57

fsih commented:
I've moved the contents of setElement into componentWillUnmount and componentDidMount like we discussed. I also fixed passing in x, y as props from the VM to position new monitors. It goes through the same method of positioning as all other monitors now, so I removed default position from the draggable.

paulkaplan commented:
Great! Looks like there are some conflicts yet to resolve also?

fsih commented:
Merged!
